From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Nicolas Iooss <nicolas.iooss_linux@m4x.org>
Date: Sat, 6 Aug 2016 11:30:56 +0200
Subject: [PATCH] {PRINTF} {WIP} include/linux/string.h: add __printf attribute
 to binary printf functions

This may be buggy because such functions do not have va_args.  Compilers
or static checkers may expect functions using __printf(..., 0) to have a
va_args somewhere in their arguments.
---
 include/linux/seq_buf.h   | 2 +-
 include/linux/string.h    | 4 ++--
 include/linux/trace_seq.h | 4 ++--
 lib/vsprintf.c            | 2 +-
 4 files changed, 6 insertions(+), 6 deletions(-)

diff --git a/include/linux/seq_buf.h b/include/linux/seq_buf.h
index 5b31c5147969..c4891dc72263 100644
--- a/include/linux/seq_buf.h
+++ b/include/linux/seq_buf.h
@@ -155,7 +155,7 @@ extern int seq_buf_hex_dump(struct seq_buf *s, const char *prefix_str,
 			    const void *buf, size_t len, bool ascii);
 
 #ifdef CONFIG_BINARY_PRINTF
-extern int
+extern __printf(2, 0) int
 seq_buf_bprintf(struct seq_buf *s, const char *fmt, const u32 *binary);
 #endif
 
diff --git a/include/linux/string.h b/include/linux/string.h
index c062c581a98b..b778b4e7b8e2 100644
--- a/include/linux/string.h
+++ b/include/linux/string.h
@@ -197,8 +197,8 @@ int __sysfs_match_string(const char * const *array, size_t n, const char *s);
 #define sysfs_match_string(_a, _s) __sysfs_match_string(_a, ARRAY_SIZE(_a), _s)
 
 #ifdef CONFIG_BINARY_PRINTF
-int vbin_printf(u32 *bin_buf, size_t size, const char *fmt, va_list args);
-int bstr_printf(char *buf, size_t size, const char *fmt, const u32 *bin_buf);
+int vbin_printf(u32 *bin_buf, size_t size, const char *fmt, va_list args) __printf(3, 0);
+int bstr_printf(char *buf, size_t size, const char *fmt, const u32 *bin_buf) __printf(3, 0);
 int bprintf(u32 *bin_buf, size_t size, const char *fmt, ...) __printf(3, 4);
 #endif
 
diff --git a/include/linux/trace_seq.h b/include/linux/trace_seq.h
index 6be92bf559fe..0bbb89e60138 100644
--- a/include/linux/trace_seq.h
+++ b/include/linux/trace_seq.h
@@ -77,7 +77,7 @@ extern __printf(2, 3)
 void trace_seq_printf(struct trace_seq *s, const char *fmt, ...);
 extern __printf(2, 0)
 void trace_seq_vprintf(struct trace_seq *s, const char *fmt, va_list args);
-extern void
+extern __printf(2, 0) void
 trace_seq_bprintf(struct trace_seq *s, const char *fmt, const u32 *binary);
 extern int trace_print_seq(struct seq_file *m, struct trace_seq *s);
 extern int trace_seq_to_user(struct trace_seq *s, char __user *ubuf,
@@ -102,7 +102,7 @@ static inline __printf(2, 3)
 void trace_seq_printf(struct trace_seq *s, const char *fmt, ...)
 {
 }
-static inline void
+static inline __printf(2, 0) void
 trace_seq_bprintf(struct trace_seq *s, const char *fmt, const u32 *binary)
 {
 }
diff --git a/lib/vsprintf.c b/lib/vsprintf.c
index be71a03c936a..22e70a5e88fd 100644
--- a/lib/vsprintf.c
+++ b/lib/vsprintf.c
@@ -2499,7 +2499,7 @@ char *pointer(const char *fmt, char *buf, char *end, void *ptr,
  * @precision: precision of a number
  * @qualifier: qualifier of a number (long, size_t, ...)
  */
-static noinline_for_stack
+static noinline_for_stack __printf(1, 0)
 int format_decode(const char *fmt, struct printf_spec *spec)
 {
 	const char *start = fmt;
-- 
